apps/frontend/src/app/Database/DataEntry.ts

```diff
@@ -24,31 +24,45 @@ export class DataEntry<
     this.key = key
     this.init = init
     this.goKey = goKey
-    if (this.database.storage.keys.includes(key)) {
-      this.data = this.toCache(init(this.database))!
-      this.set(this.database.storage.get(key))
-    } else {
-      this.data = this.toCache(init(this.database))!
-      this.setCached(this.data)
-    }
+    this.set(
+      this.database.storage.keys.includes(key)
+        ? this.getStorage()
+        : init(this.database)
+    )
+    this.data = this.get() //initializer
   }
 
   listeners: Callback<CacheValue>[] = []
   get() {
     return this.data
   }
-  validate(obj: any): StorageValue | undefined {
+  validate(obj: unknown): StorageValue | undefined {
     return obj as StorageValue | undefined
   }
   toCache(storageObj: StorageValue): CacheValue | undefined {
-    return { ...storageObj } as any as CacheValue
+    return { ...storageObj } as unknown as CacheValue
   }
   deCache(cacheObj: CacheValue): StorageValue {
-    const { ...storageObj } = cacheObj as any
-    return storageObj as any as StorageValue
+    const { ...storageObj } = cacheObj
+    return storageObj as unknown as StorageValue
+  }
+  getStorage(): StorageValue {
+    return this.database.storage.get(this.key)
   }
-  set(value: Partial<StorageValue>): boolean {
-    const old = this.data
+  set(
+    value:
+      | Partial<StorageValue>
+      | ((v: StorageValue) => Partial<StorageValue> | void)
+  ): boolean {
+    const old = this.getStorage()
+    if (typeof value === 'function') {
+      if (!old) {
+        this.trigger('invalid', value)
```

Comment:
Why do we only trigger if its a function?

Reply:
Is it ok to trigger on a function?

Reply:
It should be fine, since we are not really reading the actual values of the event listener in this context, the `invalid` events are currently only used by the import pipeline, which doesn't use the functional API.

Reply:
seems weird but i don't know enough to say anything else